src/storage/txn/actions/acquire_pessimistic_lock.rs

```diff
@@ -142,10 +142,20 @@ pub fn acquire_pessimistic_lock<S: Snapshot>(
                 None
             };
 
-        if need_load_value {
-            val = reader.get(&key, for_update_ts)?;
-        } else if need_check_existence {
-            val = reader.get_write(&key, for_update_ts)?.map(|_| vec![]);
+        if need_load_value || need_check_existence || should_not_exist {
+            let write = reader.get_write_with_commit_ts(&key, for_update_ts)?;
+            if let Some((write, commit_ts)) = write {
+                // Here `get_write_with_commit_ts` returns only the latest PUT if it exists and
+                // is not deleted. It's still ok to pass it into `check_data_constraint`.
+                if locked_with_conflict_ts.is_none() {
```

Comment:
Can you add a comment about why `locked_with_conflict_ts` cases can be skipped? I cannot think of any problem but I also want to hear how you think about it.

Reply:
Would it be a problem for this case:
1. Raise an `Insert` pessimistic lock
2. Locking is successful with conflict and the constraint check is not done
3. Response of 2 is lost, the client retry
4. The `locked_with_conflict_ts` is `Some` and the constraint check is skipped unexpectedly

Reply:
It acquires the lock and lets the client (TiDB) retry the statement in this case. TiDB will find the key already exists when it retries the statement.
But actually I want to adjust this behavior later: When `should_not_exist` is set and the key exists and the latest version is newer than for_update_ts, return write conflict error even `allow_lock_with_conflict` is set. I think in most cases, when the statement retries, it's likely that it still want to insert the same key and it should still fail.

Reply:
> Would it be a problem for this case:
> 
> 1. Raise an `Insert` pessimistic lock
> 2. Locking is successful with conflict and the constraint check is not done
> 3. Response of 2 is lost, the client retry
> 4. The `locked_with_conflict_ts` is `Some` and the constraint check is skipped unexpectedly

 In the current implementation (which I'm planning to adjust later), the retried request in the 4th step will still produce a `locked_with_conflict` result, and TiDB will still retry. It will notice the key already exists when retrying the statement and abort the statement then, at which time the lock will be `pessimistic_rollback`-ed

Reply:
Got it. So the correctness depends on that the returned pessimistic lock result must contain "value existing" information, maybe we could add this in the comment too in case we forget to return the required information?

Reply:
The `Value` field is required in [`PessimisticLockKeyResult::LockedWithConflict`](https://github.com/tikv/tikv/blob/e1467c56a445d36a8fd8642f9467a0b18fbb8203/src/storage/types.rs#L174). I think it's just fine for locked_with_conflict cases? I'm actually not very sure where you prefer to add the comment you said

Reply:
Ok, it's fine.